firewood/src/merkle/proof.rs

```diff
@@ -258,261 +257,146 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
+            let child_node = match &parent_node_ref.inner() {
                 #[allow(clippy::indexing_slicing)]
-                NodeType::Branch(n) => match n.chd()[branch_index] {
+                NodeType::Branch(n) => match n.chd()[child_index] {
                     // If the child already resolved, then use the existing node.
-                    Some(node) => {
-                        chd_ptr = node;
-                    }
+                    Some(node) => merkle.get_node(node)?,
                     None => {
+                        let child_node = decode_subproof(merkle, proofs_map, &child_hash)?;
+
                         // insert the leaf to the empty slot
                         #[allow(clippy::unwrap_used)]
-                        u_ref
-                            .write(|u| {
-                                let uu = u.inner_mut().as_branch_mut().unwrap();
+                        parent_node_ref
+                            .write(|node| {
                                 #[allow(clippy::indexing_slicing)]
-                                (uu.chd_mut()[branch_index] = Some(chd_ptr));
+                                let node = node.inner_mut().as_branch_mut().unwrap();
+                                node.chd_mut()[child_index] = Some(child_node.as_ptr());
                             })
                             .unwrap();
+
+                        child_node
                     }
                 },
 
-                NodeType::Extension(_) => {
-                    // If the child already resolved, then use the existing node.
-                    #[allow(clippy::unwrap_used)]
-                    let node = u_ref.inner().as_extension().unwrap().chd();
-
-                    #[allow(clippy::unwrap_used)]
-                    if node.is_null() {
-                        u_ref
-                            .write(|u| {
-                                let uu = u.inner_mut().as_extension_mut().unwrap();
-                                *uu.chd_mut() = chd_ptr;
-                            })
-                            .unwrap();
-                    } else {
-                        chd_ptr = node;
-                    }
+                #[allow(clippy::unwrap_used)]
+                NodeType::Extension(n) if n.chd().is_null() => {
+                    let child_node = decode_subproof(merkle, proofs_map, &child_hash)?;
+
+                    parent_node_ref
+                        .write(|node| {
+                            let node = node.inner_mut().as_extension_mut().unwrap();
+                            *node.chd_mut() = child_node.as_ptr();
+                        })
+                        .unwrap();
+
+                    child_node
                 }
 
+                NodeType::Extension(n) => merkle.get_node(n.chd())?,
+
                 // We should not hit a leaf node as a parent.
                 _ => return Err(ProofError::InvalidNode(MerkleError::ParentLeafBranch)),
```

Comment:
We should not panic there. I think it would make sense to return a "database is corrupt" error here, but if we panic, that doesn't give the database user the opportunity to gracefully handle anything else they have in flight. It should be the user's decision whether or not to panic.